### ISSUE TITLE
ATC memory issue after upgrade 2.6.0

### ISSUE THREAD

**Issue by ahelal** [concourse/concourse#879]
Hi there!

# Bug Report

After upgrading to 2.6.0 the ATC node started to fail after a while 6 hours or so. It keeps consuming memory till it fails. 

Prior to upgrade everything was fine version 2.5.1

I attach a memory graph 
![atc-memory](https://cloud.githubusercontent.com/assets/4069495/21957605/20d15862-da9a-11e6-86a4-557c4151f796.png)

* Concourse version: 2.6.0 and 2.7.0
* Deployment type: binary
* Infrastructure/IaaS: AWS t2.medium
* Did this used to work? yes
* Linux : Ubuntu 16.04 Kernel 4.4.0-62 

Only work around so far is to restart concourse web twice a day

**Comment by wieslaw-gat** [comment]
I've also encountered this problem, `concourse web` keeps hogging more and more memory - the usage goes up to ~2GB after which I have to restart it before oomkiller reaps it.

This started happening after the upgrade to 2.6.0, earlier no such problems appeared.

**Comment by ahelal** [comment]
I also attached the heap from pprof might be helpful https://gist.github.com/ahelal/5af9bd99f430ea88e89e7c76d2ceb95e

**Comment by arwineap** [comment]
Not a whole lot to add, but I did verify that my install also started gobbling memory on the web node.

http://i.imgur.com/qu5cPhn.png

**Comment by schubert** [comment]
FTW since we saw this and reported it in #892 we believe this event correlated with a job failing on a worker at the same time with the following failure:

```
! DETAIL:  could not fork new process for connection: Resource temporarily unavailable
```

Which I completely speculate could be caused by exhausting file descriptors on the worker which could be caused by the many stdout/stderr handlers open and perhaps not being cleaned up and therefore exhausting FD's on the worker while simultaneously holding onto memory over on the web VM leading to the memory exhaustion.

Or it was just a coincidence but I thought it would be worth mentioning the correlation.

cc @tom-meyer

**Comment by dsharp-pivotal** [comment]
Since it's not mentioned so far here: the leak seems to be in the TSA process that runs alongside the ATC on the same node.

I noticed an uptick in memory usage on one of our web nodes around 16:00 PST today, so I looked at the TSA logs around then.

I've attached around 5k lines of logs. There do seem to be some differences in the logs around 16:00, though nothing I can make sense of.

Here's an interesting snippet from around the time:
```
{"timestamp":"1485419197.817038059","source":"tsa","message":"tsa.connection.handshake-failed","log_level":1,"data":{"error":"EOF","remote":"10.0.2.155:1025","session":"47"}}
{"timestamp":"1485438864.030432701","source":"tsa","message":"tsa.connection.handshake-failed","log_level":1,"data":{"error":"EOF","remote":"10.0.2.186:47367","session":"2"}}
{"timestamp":"1485444172.238133192","source":"tsa","message":"tsa.connection.handshake-failed","log_level":1,"data":{"error":"EOF","remote":"10.0.2.155:37363","session":"49"}}
{"timestamp":"1485447208.702469826","source":"tsa","message":"tsa.connection.handshake-failed","log_level":1,"data":{"error":"ssh: invalid packet length, packet too large","remote":"10.0.2.84:14584","session":"50"}}
{"timestamp":"1485474638.292750835","source":"tsa","message":"tsa.connection.channel.channel-request","log_level":1,"data":{"remote":"10.0.16.31:39028","session":"51.1","type":"exec"}}
{"timestamp":"1485474638.293791294","source":"tsa","message":"tsa.connection.channel.register-worker.register.start","log_level":1,"data":{"remote":"10.0.16.31:39028","session":"51.1.1.5","worker-address":"10.0.16.31:7777","worker-platform":"linux","
worker-tags":""}}
Thu Jan 26 15:50:38 PST 2017
{"timestamp":"1485474638.294978857","source":"tsa","message":"tsa.connection.channel.register-worker.register.reached-worker","log_level":0,"data":{"remote":"10.0.16.31:39028","session":"51.1.1.5","took":"1.088562ms"}}
{"timestamp":"1485474638.306630373","source":"tsa","message":"tsa.connection.channel.register-worker.register.done","log_level":1,"data":{"remote":"10.0.16.31:39028","session":"51.1.1.5","worker-address":"10.0.16.31:7777","worker-platform":"linux","worker-tags":""}}
{"timestamp":"1485474646.315000296","source":"tsa","message":"tsa.connection.keepalive","log_level":1,"data":{"remote":"10.0.16.31:39028","session":"51","type":"keepalive@openssh.com"}}
{"timestamp":"1485474654.323520660","source":"tsa","message":"tsa.connection.keepalive","log_level":1,"data":{"remote":"10.0.16.31:39028","session":"51","type":"keepalive@openssh.com"}}
{"timestamp":"1485474662.332080603","source":"tsa","message":"tsa.connection.keepalive","log_level":1,"data":{"remote":"10.0.16.31:39028","session":"51","type":"keepalive@openssh.com"}}
{"timestamp":"1485474668.306895733","source":"tsa","message":"tsa.connection.channel.register-worker.heartbeat.start","log_level":1,"data":{"remote":"10.0.16.31:39028","session":"51.1.1.6","worker-address":"10.0.16.31:7777","worker-platform":"linux","worker-tags":""}}
{"timestamp":"1485474668.308068752","source":"tsa","message":"tsa.connection.channel.register-worker.heartbeat.reached-worker","log_level":0,"data":{"remote":"10.0.16.31:39028","session":"51.1.1.6","took":"1.044977ms"}}
{"timestamp":"1485474668.318796873","source":"tsa","message":"tsa.connection.channel.register-worker.heartbeat.done","log_level":1,"data":{"remote":"10.0.16.31:39028","session":"51.1.1.6","worker-address":"10.0.16.31:7777","worker-platform":"linux","worker-tags":""}}
```

This looks significantly different than the steady stream of "tsa.connection.tcpip-forward.forward-conn" messages filling the rest of the logs.

[tsa.stdout.log.20170126.circa.1600PST.leak.gz](https://github.com/concourse/concourse/files/734314/tsa.stdout.log.20170126.circa.1600PST.leak.gz)

**Comment by artem-zinnatullin** [comment]
Confirming that OOM was introduced in 2.6.0, falling back to 2.5.1 fixes the problem.

**Comment by ahelal** [comment]
I upgrade to 2.7 issue is still valid.
Did anybody find a work around ?  Concourse is not usable like this reboot twice a day.

**Comment by voelzmo** [comment]
which stemcell are you running? stemcells >3312.12 have a known memory leak in the kernel: https://github.com/cloudfoundry/bosh/releases/tag/stable-3312.17

**Comment by ahelal** [comment]
@voelzmo it's binary not using bosh

**Comment by markstgodard** [comment]
Which Linux kernel version are you running the binaries on?  Linux 4.4.0_59 is known to have an issue and I had to revert to 4.4.0_57

**Comment by ahelal** [comment]
@markstgodard Ubuntu 16.04 Kernel 4.4.0-62

So you think downgrading can help ?

**Comment by markstgodard** [comment]
It helped in my case, but I still see memory growing over time, so it seems like there might still be an issue

Before downgrading kernel, atc was getting OOM everyday or so with light usage at the time. 

I am not sure if the memory issue in 4.4.0_59 had been addressed in 62 but something to consider. The Linux issue link should be in Concourse slack if you search for that kernel number

Good luck

**Comment by voelzmo** [comment]
See https://bugs.launchpad.net/ubuntu/+source/linux/+bug/1655842
either downgrade to <4.4.0-59 or upgrade to >=4.4.0-63

**Comment by ahelal** [comment]
@voelzmo 63 is not out yet for Ubuntu 16.04 will wait and report if it works

**Comment by ahelal** [comment]
Did not work. I upgrade to **63** using apt proposed source. And we are still facing the same issue.

**Comment by arwineap** [comment]
The kernel OOM from ubuntu and the atc growing memory usage are two separate issues.

I am seeing the issues on the -57 kernel

**Comment by rnurgaliyev** [comment]
Issue persists on 16.04 with 4.4.0-65 kernel. Doesn't look like kernel issue anymore.

**Comment by billimek** [comment]
Possibly related to the ATC always logging at debug level? https://github.com/concourse/concourse/issues/950

**Comment by billimek** [comment]
Confirming similar memory patterns running concourse (db, web, worker) via docker and seeing memory linearly grow and then 'reset' every time the concourse web container would crash and restart.  This is a view of a 10-day period running v 2.7.0:

![](http://i.imgur.com/jHPaWUg.png)

**Comment by ahelal** [comment]
@vito is it possible to address this issue in the near future ? as it is a growing pain? 
I am willing to help with whatever is possible to make it easier for you guys.

**Comment by vito** [comment]
@ahelal Sorry for the low visibility on this, I've not been dividing my time well between GH issues and engineering lately (just getting back into engineering/pairing so focusing on that). Hoping to solve that general problem by onboarding the whole team on GH issue management and making it a regular part of our day-to-day, rather than mainly me.

This definitely looks pretty serious, though it's odd that we don't observe this on our own instance.

Thanks all for the metrics screenshots - that really helps confirm the issue.

I'm curious why we're not seeing this. It could be a difference in our pipelines (maybe you use `resource_types` and we don't for example). I wonder if you could pause certain pipelines and see if the memory growth rate plateaus or slows down? It'd also be worth determining if it's a leak in the TSA or a leak in the ATC, which will be hard with the binary distribution as they're fused into one executable. But maybe you could run the individual processes and observe that? The flags should be pretty similar, just with separate binaries. You'd have to `git checkout v2.7.0` in the `concourse` repo, `export GOPATH=$PWD`, and `go install github.com/concourse/atc` and `go install github.com/concourse/tsa` for that.

**Comment by artem-zinnatullin** [comment]
Since I've faced this issue on a setup with only one pipeline on the first day I've tried to integrate Concourse with our project, I think it can be used as pretty isolated example:

```yml
resources:
- name: source-code
  type: git
  source:
    …

jobs:
- name: pull-requests
  plan:
  - get: source-code
    trigger: true
  - task: build
    privileged: true
    config:
      platform: linux
      image_resource:
        type: docker-image
        source: 
          repository: amidos/dcind
      inputs:
      - name: source-code

      run:
        path: sh
        args:
          - -ec
          - |
            source /docker-lib.sh
            start_docker
            env
            apk --update --no-cache add bash
            bash source-code/ci.sh
```

TL;TR: I'm pulling source code from `git`, then I run Docker container in privileged mode from image with `docker` inside of it so I could run Docker during the build and that's basically it.

In my case on machine with 16 gb of memory that runs Concourse Web and Concourse Worker only I could only build project three or four times and then it'll crash with OOM.

If that helps: during the build, inner docker pulls image that weights about 8 gb and based on `ubuntu:16.04`.

**Comment by vito** [comment]
@artem-zinnatullin That sounds suspiciously close to the kernel issue mentioned earlier in the thread, which is different. Are you seeing a leak in the `web` process?

**Comment by artem-zinnatullin** [comment]
@vito well, the problem is that we build 2 projects about 80-100 times every day using same base image and Jenkins as a CI and do not see this issue…

As far as I remember, `concourse-web` was crashing, but I think I've also seen `concourse-worker` failing few times. It would take some time to allocate machine, setup Concourse and repeat the process, but I can do it if noone will post another isolated example.

**Comment by markstgodard** [comment]
I run concourse deployment on GCP using just the binaries and some monit scripts (not using docker compose or bosh) and I do see memory on the atc grow over time.

**Comment by dsharp-pivotal** [comment]
@vito As I said in https://github.com/concourse/concourse/issues/879#issuecomment-275577903, the leak is definitely in the TSA.

We only see the leak in one of our Concourses. One difference I was wondering about is if it has anything to do with our externally-registered workers on that instance.

**Comment by rnurgaliyev** [comment]
@vito it is actually process consuming and not freeing memory, definitely not kernel OOM feature bug.

**Comment by voelzmo** [comment]
We're also seeing this on stemcell `3363.9` and concourse release `2.7.0`. It takes a day for TSA to consume ~75% of 16GB Ram on an m4.xlarge instance. Workaround for now is restarting TSA once a day ;(

**Comment by markstgodard** [comment]
I am attaching pprof dump of my atc,

Concourse 2.6.0 running the straight binaries on Linux amd64, kernel:  4.4.0-57-generic
I was experiencing the issue w/ that linux kernel bug, but downgraded and that addressed that issue. But still what looks like a memory leak:

![atc-mem](https://cloud.githubusercontent.com/assets/405082/24324005/8043e4ca-114c-11e7-9b59-667586ab4561.png)

**Comment by shinji62** [comment]
I am wondering if we should not defer the listener.Close() here https://github.com/concourse/tsa/blob/4f20e8094914a985928095eaa6921cc14d611b69/tsacmd/server.go#L583

**Comment by rnurgaliyev** [comment]
It seems something is fixed in v2.7.1-rc.36. I wonder if it was something to do with TSA flooding log issue? 
![image](https://cloud.githubusercontent.com/assets/22738239/24542139/bce30732-15fa-11e7-9f52-ae36c741185b.png)
